google-cloud-firestore/lib/google/cloud/firestore/collection.rb

```diff
@@ -0,0 +1,663 @@
+          # @return [String] collection identifier.
+          def collection_id
+            path.split("/").last
+          end
+
+          ##
+          # A string representing the path of the collection, relative to the
+          # document root of the database.
+          #
+          # @return [String] collection path.
+          def collection_path
+            path.split("/", 6).last
+          end
+
+          ##
+          # A string representing the full path of the collection resource.
+          #
+          # @return [String] collection resource path.
+          def path
+            @path
+          end
+
+          ##
+          # The document reference or database the collection reference belongs
+          # to. If the collection is a root collection, it will return the
+          # database object. If the collection is nested under a document, it
+          # will return the document reference object.
+          #
+          # @return [Database, Collection::Reference] parent object.
+          #
+          # @example Returns database object for root collections:
+          #   require "google/cloud/firestore"
+          #
+          #   firestore = Google::Cloud::Firestore.new
+          #
+          #   # Get a collection reference
+          #   cities_col = firestore.col "cities"
+          #
+          #   # Get the document's parent collection
+          #   database = cities_col.parent
+          #
+          # @example Returns document object for nested collections:
+          #   require "google/cloud/firestore"
+          #
+          #   firestore = Google::Cloud::Firestore.new
+          #
+          #   # Get a collection reference
+          #   precincts_ref = firestore.col "cities/NYC/precincts"
+          #
+          #   # Get the document's parent collection
+          #   nyc_ref = precincts_ref.parent
+          #
+          def parent
+            if collection_path.include? "/"
+              return Document.from_path parent_path, context
+            end
+            return context.database if context.respond_to? :database
+            context
+          end
+
+          ##
+          # Retrieves a list of documents.
+          #
+          # @yield [documents] The block for accessing the documents with data.
+          # @yieldparam [Document::Snapshot] document A document with data.
+          #
+          # @return [Enumerator<Document::Snapshot>] documents with data list.
+          #
+          # @example
+          #   require "google/cloud/firestore"
+          #
+          #   firestore = Google::Cloud::Firestore.new
+          #
+          #   # Get a collection reference
+          #   cities_col = firestore.col "cities"
+          #
+          #   # Get and print all city documents
+          #   cities_col.docs do |city|
+          #     puts "#{city.document_id} has #{city[:population]} residents."
+          #   end
+          #
+          def docs &block
+            query.run(&block)
```

Comment:
Can we make CollectionReference inherit from Query, or is that not a thing in Ruby? This would greatly reduce the code and comments in this class.

Reply:
We can, but Rubyists generally prefer composition over inheritance. That is why both Collection::Reference and Query support the same interface. The implementation mechanism isn't as important. This is known as "Duck Typing".

I believe there are advantages to both the implementation and documentation for keeping them separate and not using inheritance here.

Reply:
Having the same documentation here generates a lot of code duplication. I would rather we concentrate this documentation in one place.

FWIW, every other client uses inheritance to solve this problem.